### PR TITLE
Added numerous bugfixes for cloning and deferred planning.

### PR DESCRIPTION
**Pull request by psigen**
* Deferred planning now consistently returns trollius.futures.Future
* Fixed bug in robot PlanWrapper that caused deferred planning to terminate early.
* Cloned() references are now explicitly passed their clone environment.
* .Cloned() helper method added to environments created by Clone(env)
* Existing clone references consolidated to minimize Cloned() lookups.
* Added usage of CopyTrajectory helper instead of direct Clone calls where applicable.

### CODE REVIEW

File: src/prpy/base/manipulator.py [hunk 1]
```diff
@@ -141,19 +141,25 @@ def SetVelocityLimits(self, velocity_limits, min_accel_time):
         self.GetRobot().SetDOFAccelerationLimits(or_accel_limits)
 
     def _PlanWrapper(self, planning_method, args, kw_args):
-        from prpy.clone import Clone, Cloned
+        from prpy.clone import Clone
         robot = self.GetRobot()
-        with Clone(robot.GetEnv()):
-            Cloned(self).SetActive()
+        with Clone(robot.GetEnv()) as cloned_env:
```
Comment (mkoval) [inline src/prpy/base/manipulator.py]: I think we should pass `lock=True` here.

File: src/prpy/base/mico.py [hunk 2]
```diff
@@ -139,21 +139,23 @@ def PlanToNamedConfiguration(self, name, execute=True, **kw_args):
         robot = self.GetRobot()
         saved_dof_indices, saved_dof_values = robot.configurations.get_configuration(name)
 
-        with Clone(robot.GetEnv()):
-            Cloned(self).SetActive()
-            arm_dof_indices = Cloned(robot).GetActiveDOFIndices()
-            arm_dof_values = Cloned(robot).GetActiveDOFValues()
+        with Clone(robot.GetEnv()) as cloned_env:
```
Comment (mkoval) [inline src/prpy/base/mico.py]: There should also be a `lock=True` here.

File: src/prpy/base/manipulator.py [hunk 3]
```diff
@@ -141,19 +141,25 @@ def SetVelocityLimits(self, velocity_limits, min_accel_time):
...
+            cloned_env.Cloned(self).SetActive()
```
Comment (mkoval) [inline src/prpy/base/manipulator.py]: What was wrong with the previous `Cloned(self).SetActive()`. There are no threads here, so it should be unambiguous. Is this just a matter of preference?

File: src/prpy/base/mobilebase.py [hunk 4]
```diff
@@ -145,27 +145,29 @@ def DriveStraightUntilForce(self, direction, velocity=0.1, force_threshold=3.0,
...
-        from prpy.clone import Clone, Cloned
-        
         robot = self.robot
-        with Clone(robot.GetEnv()):
-            Cloned(robot).SetActiveDOFs([], 
-                                        affine = openravepy.DOFAffine.X |
-                                        openravepy.DOFAffine.Y | openravepy.DOFAffine.RotationAxis)
-
-            cloned_traj = planning_method(Cloned(robot), *args, **kw_args);
+        with Clone(robot.GetEnv()) as cloned_env:
```
Comment (mkoval) [inline src/prpy/base/mobilebase.py]: There should also be a `lock=True` here.

File: src/prpy/base/robot.py [hunk 5]
```diff
@@ -317,14 +320,26 @@ def postprocess_trajectory(traj, kw_args):
...
         # Return either the trajectory result or a future to the result.
         if 'defer' in kw_args and kw_args['defer'] is True:
-            from trollius.executor import get_default_executor
-            executor = kw_args.get('executor') or get_default_executor()
-            return executor.submit(defer_trajectory, result, kw_args)
+            import trollius
+
+            # Perform postprocessing on a future trajectory.
+            @trollius.coroutine
+            def defer_trajectory(traj_future, kw_args):
```
Comment (mkoval) [inline src/prpy/base/robot.py]: I thought we mostly moved away from Trollius coroutines to Trollius futures. Is this an artifact of this branch being out of date? Or are we still using coroutines in some places?

**Comment by mkoval** [comment]
This looks good to me, other than those missing locks.

I also wish you'd write documentation for #48. I'm still a bit confused myself by the way we ended up using Trollius. :smile:

File: src/prpy/base/manipulator.py (hunk 3 above)
Comment (psigen) [inline src/prpy/base/manipulator.py]: Nothing wrong here, just making it consistent.

**Pull request merged**